---
title: "TITAN Tutorial"
always_allow_html: yes  
output:
  github_document:
    toc: true
    toc_depth: 5
vignette: |
  %\VignetteIndexEntry{TITAN Tutorial} %\VignetteEngine{knitr::rmarkdown} %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  cache=T
)
library(knitr)
```


## What is Titan?

  TITAN (Topic Inference of Transciptionally Associated Networks) is an R package that runs Topic Modeling on scRNA-seq data. Topic modeling in the traditional sense has been used in text mining to link words and documents to topics. Topics in the traditional sense  can be thought of as themes, such as categories in news article inference (science, politics, etc). TITAN applies topic modeling to scRNA-seq to find latent transcriptional topics that link genes to topics and cells to topics. TITAN utalizes a nlp alghorithm called Latent Dirchlet Allocation (LDA) that maximizes these two probability distributions that show 1) the probability a gene is assigned to a topic and 2) the probability a cell is assigned to a topic. 


## Object Setup

## Package loading and installation 

```{r, eval = F}
install.packages("devtools")
devtools::install_github('JuliusCampbell/TITAN')
```


```{r message = F, warning=F}
library(TITAN)
library(Seurat)
library(tidyverse)
```

In its current state, TITAN is built to work off of the R package `Seurat`. The first step in TITAN is to create a Seurat object containing your scRNA-seq data. This tutorial assumes that general QC has been run and poorly performing cells have been filtered out. Directions for that can be found in the vingette found [here](https://satijalab.org/seurat/v3.1/pbmc3k_tutorial.html). We also understand that there are other formats of single cell objects, such as SingleCellExpiriment. Seurat has documentation on conversion to SeuratObject from different types of objects found [here](https://satijalab.org/seurat/v3.1/conversion_vignette.html). Once you've done this, you can begin using TITAN. To begin building the model, the user has two options. 

For this tutorial, the Seurat object we will be using can be downloaded as an RDS object from [thislink](https://www.dropbox.com/s/sjzt0720ijj0ycs/PEPE_T47D.rds?dl=0). This is a cancer cell line hormonal treatment time course. The dataset contains T47D single cells, some treated with estrogen, some treated with progesterone, and some treated with both estrogen and progesterone (which we refer to as PEPE treatment).

## Building a Model

The first is running a single LDA model on the data. This process has two steps. First is to normalize the data. We do this using Seurats normalize data function with the added recommendation that one uses the centered log-ratio (CLR) transformation due to the varying levels of gene expression seen per cell and topic modeling's preference for genes that have high counts. 

### Single topic
```{r eval = F}
SeuratObj <- readRDS("PEPE_T47D.rds")
LDA_model <- runLDA(SeuratObj, ntopics = 50, normalizationMethod = "CLR", seed.number = 8)

```

This will return an LDA model with 50 topics, created using the default alpha and beta values, as well as the default values for a number of other parameters. More info about these parameters can be found at `?runLDA`.min Depending on the size of your data, this function can take multiple hours to run, so it is important to make sure you have the necessary computational resources. 

### Multiple Topics

The second option is to use the parallel option of the `runLDA` function. This is for users that may not know how many topics they want in their model. This function does the same thing as above, but a number of times in parallel, each time using a different number of topics. Now the `ntopics` argument takes a vector of numbers and the function will produce a new model for every number of topics in the list.

```{r eval = F}
runLDA(SeuratObj, ntopics = seq(5, 50, by=5), parallel = TRUE, outDir = "output_directory", cores = 10)

```

The above command will run 10 instances of the `runLDA` function starting with only using 5 topics, going up to using 50 topics, stepping up by 5 each time. The 10 instances are run at the same time in parallel. The models will be output as `.rds` files into the given output directory. 

## Evaluating Models

### Elbow Plot nTopics

Now the task is to determine which of these models is the ideal one to use in downstream analysis. To do this we have developed our own version of an Elbow plot. This elbow plot looks at the perplexity of each model, a measure how well the model fits the data while also giving models that use less parameters a higher score. The Y axis of the plot is the RPC (rate of perplexity change), which looks at how much the perplexity changes as the topic number increases. This method is based off [this](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-16-S13-S8) paper by Zhao et al.

```{r eval = F}
LDAelbowPlot("output_directory", SeuratObj)

```

![Elbow Plot](../data/Elbowplot_T47D_PEPE.png){ width=40% }


## Analyzing Models

Now that we have built a model and determine that it does a good job of capturing our data, we can analyze the model. For the rest of the vignette, we will be using a model with 20 topics that was built on the dataset. Note that the LDA model is an indeterministic model, meaning the initial starting points will effet the output of the model. For this reason, setting a seed is important and the results above may be very very similar but not the exact same as what was recapitulated in the paper. For this reason, we will be showing an example of the results of our titan run from the manuscript which was generated with an internal seed. That file can be loaded [here](https://www.dropbox.com/s/t80us62cg1zm1vu/Model_PEPE_T47D_20T_CLR_5000Variable_M10.rds?dl=0).


### Load Data 

First, read in the model RDS you just generated.

```{r, eval = F, include = T}
top_model <- readRDS("../data/Model_PEPE_T47D_20T_CLR_5000Variable_M10.rds")
```

```{r, eval = T, include = F}

SeuratObj <- readRDS("~/Box Sync/Alex_Data/MCF7/Package/PEPE_T47D.rds")
top_model <- readRDS("~/Box Sync/Alex_Data/MCF7/PEPE/Model_PEPE_T47D_20T_CLR_5000Variable_M10.rds")
MCF7_top_model <- readRDS("~/Box Sync/Alex_Data/MCF7/MCF7_New/Model_MCF7_PEPE_20T_CLR_5000Variable_M10.rds")
```

The first step in the analysis pipeline is to add the topic information to the original Seurat object. This allows us to employ many of Seurat's built in plot functions to visualize the model.

```{r}

SeuratObj <- addTopicsToSeuratObject(model = top_model, Object = SeuratObj)
```


### Gene-Topic Distribution

Now that the topic information has been stored within the Seurat object, we can begin analyzing the contents of the topics. There are two distrubutions that LDA outputs. The first being the assignments of genes in each topic, using the scores of each gene per topic we can look at which genes contribute most per topic. The first function here, `GeneScores` outputs a matrix with a table of the scores for each gene. Sorting by topic, one can see the disribution of genes and which genes contribute most to each topic. Below is an example of showing only the top topic 1 genes and subsequent scores of said for the first 10 topics. 


```{r}

GeneDistrubition <- GeneScores(top_model)

### head and sort the matrix by genes most contributing to Topic 1
kable(head(GeneDistrubition[order(GeneDistrubition[,"Topic_1"], decreasing = T),],n = 10))

```


An alternative to this matrix is to get a matrix of the top n genes in each topic. This can be done with `TopTopicScores`, which sorts the gene-topic matrix and outputs a matrix list of n genes for eaech topic which can be inputted into your favorite GO program.These genes can be used to connect each topic to a certain gene network or ontology. 

``` {r,}

TopicGenes <- TopTopicGenes(top_model, ngenes = 50)
kable(head(TopicGenes[,1:10], 10))
kable(head(TopicGenes[,11:20], 10))
```

### Cell-Topic Distribution

The second distrubtion that TITAN outputs is the cell-topic distrubition which outputs the contribution each cell belongs to a topic, calculated from the gene scores. The `GetTopics` function outputs a scaled cell-topic matrix. Alternatively, the `addTopicsToSeuratObject` function added this cell-topic matrix to our metadata files so if that has been run it can also be done by wrangling the meta.data in your Seurat object.

``` {r,}

## Using the get topics function
LDA_topics <- GetTopics(top_model, SeuratObj)
kable(head(LDA_topics[,1:10]))
kable(head(LDA_topics[,11:20]))

## Using the meta data columns (must run addTopicsToSeruatObject() first)
LDA_topics <- SeuratObj@meta.data %>% select(starts_with("Topic"))
kable(head(LDA_topics[,1:10]))
kable(head(LDA_topics[,11:20]))
```


## Visualizing Topics 

### Heatmap

Although, the cell-topic matrix contains valuable infomration about each cells contribution to each topic. It is often hard to parse. For this reason, it is nice to visualize which cells contribute to each topic and how each topic compares to eachother.To do this, we recommend visualizing the cell-topic matrix in heatmap form. The goal is to see which cells express which topics. The `heatmap` function allows one to plot the topics across a heatmap and sort the cells by a column of interest. In our case, we are wondering whihch topics are associated with each treatment so we will be annotating the function with our treatment column, but one could use any other metadata column

```{r, fig.width=11, fig.height=8.5}

HeatmapTopic(Object = SeuratObj,
        topics =  LDA_topics,
        AnnoVector = SeuratObj@meta.data$hash.ID,
        AnnoName = "Time Point")


```

You can also cluster the topics of the heatmap so that topics with similar expression patterns across cells are grouped together.

```{r, fig.width=11, fig.height=8.5}

HeatmapTopic(Object = SeuratObj,
             topics =  LDA_topics,
             AnnoVector = SeuratObj@meta.data$hash.ID,
             AnnoName = "Time Point",
             clusterTopics = T)

```

In both of the above cases, the rows of the heatmap were sorted by time point. They can be sorted by any metadata column in the Seurat object as well, such as cluster.

```{r, fig.width=11, fig.height=8.5}

HeatmapTopic(Object = SeuratObj,
             topics =   LDA_topics, 
             AnnoVector = SeuratObj$seurat_clusters, 
             AnnoName = "Cluster")


```


### On UMAP

Another way to visualize the expression of the topics is through Seurat's FeaturePlot function. This plots the cells using the UMAP embeddings within the Seurat object, and then colors the cells based on their expression of the given topic.

```{r, fig.width=11, fig.height=8.5}

FeaturePlot(SeuratObj, 
            pt.size = 0.01, 
            features  = "Topic_1", 
            min.cutoff = 'q1')
FeaturePlot(SeuratObj, 
            pt.size = 0.01, 
            features  = paste("Topic", 13:20, sep = "_"), 
            ncol = 4, 
            min.cutoff = 'q1')

```


## Plotting two LDA topics against eachother

### By Factor

Further, we find it also useful to plot two LDA topics against eachother when trying to compare how cells react to different tracriptional topics. An example being in our T47D expriment, we were interested in how the progestorne-specific topic (Topic 1) and an estrogen-specific topic (Topic 17) segregate out cells by treatment. To do this, we can plot the scores of the two topics using the `DimPlot` function in Seurat. When running `addTopicsToSeruatObject`, the function creates a DimReducObject under the name "lda" which can be used to plot the two against eachother. 

```{r, fig.width=8, fig.height=6}

DimPlot(SeuratObj, 
        pt.size = 0.01, 
        reduction = "lda", 
        dims = c(17,1),
        group.by = "hash.ID")    
```

### Visualizing Expression within network gradients

Further, if one wanted to visualize the expression of a gene across two topic gradients one could also utalize the `FeaturePlot` function using the "lda" reduction as shown below. 
``` {r, fig.width=8, fig.height=6}

FeaturePlot(SeuratObj, 
        pt.size = 0.01, 
        reduction = "lda", 
        dims = c(17,1),
        features = "AZGP1",
        min.cutoff = 'q1')        

```

Another example of how one could look at the data would be to see how two topics that have similar patterns differ from eachother. If one looks back at the heatmap, topics 11 and topics 17 both showed high expression at the late estrogen timepoint, yet when ploted against eachother one can see a start pattern of cells that either undergo one topic or the other, but uncommon for both topics to be high in the same cell.

``` {r, fig.width=8, fig.height=6}

DimPlot(SeuratObj, 
        pt.size = 0.01, 
        reduction = "lda", 
        dims = c(11,17),
        group.by = "hash.ID")        

```



### Imputing Topics 

In addition, because of the way that topic modeling works, comparisons of gene sets that are enriched in topics from different expiriments can be compared from dataset to dataset by looking at the expression of the top genes within the topic. We do this by using Seurat's `AddModuleScore` function which uses the method established by [Teirosh et al](https://science.sciencemag.org/content/352/6282/189). Here is an example of the same treatment timecourse in the MCF7 cell line (Same treatments P, E & EP). This allows one to visualize how the MCF7 topics are seen under the same conditions in the t47D cell line, One finding of note is that Topic 9 in the MCF7 samples was the Estrogen specific topic. Also interestingly, topic modeling by 20 topics for MCF7 did not locate a progestorone specific topic. 

### Load Data

First, read in the MCF7 PEPE model RDS which can be found [here](https://www.dropbox.com/s/procxdnedx2m6or/Model_MCF7_PEPE_20T_CLR_5000Variable_M10.rds?dl=0).

```{r, eval = F, }
top_model <- readRDS("../data/Model_PEPE_T47D_20T_CLR_5000Variable_M10.rds")
```

```{r, fig.width=11, fig.height=8.5, warning=FALSE}
SeuratObj <- ImputeAndAddTopics(SeuratObj, MCF7_top_model, TopicPrefix = "MCF7Imputed_Topic")
HeatmapTopic(Object = SeuratObj,
             topics =   Embeddings(SeuratObj, "imputedLDA"),
             AnnoVector = SeuratObj@meta.data$hash.ID,
             AnnoName = "Cluster")



```





<!-- ### Using a SingleCellExperiment Object -->

<!-- Building the model is ran the same way, -->

<!-- ```{r} -->
<!-- SCE <- readRDS("../data/PEPE_T47D_SCE.rds") -->
<!-- ``` -->

<!-- Running a single model, -->

<!-- ```{r eval = F} -->
<!-- LDA_model <- runLDA(SCE, ntopics = 50) -->

<!-- ``` -->

<!-- Running multiple models, -->

<!-- ```{r eval = F} -->
<!-- parallel_runLDA(SCE, "output_directory", top_start = 5, top_end = 50, step = 5) -->

<!-- ``` -->

<!-- Checking the number of topics through an Elbow plot -->

<!-- ```{r eval = F} -->
<!-- LDAelbowPlot("output_directory", SCE) -->

<!-- ``` -->

<!-- ![Elbow Plot](../data/Elbowplot_T47D_PEPE_SCE.png){ width=40% } -->

<!-- Loading in the pre-built model using the same number of topics as before, -->

<!-- ```{r} -->
<!-- SCE_model <- readRDS("../data/Model_SCE_PEPE_T47D_20T.rds") -->
<!-- ``` -->

<!-- This model will be slightly different than the one ran using the Seurat object. This is due to the SingleCellExperiment workflow using a slightly different feature space. However, the results remain mostly the same. -->

<!-- ```{r} -->
<!-- SCE <- addTopicsToSCE(SCE_model, SCE) -->

<!-- ``` -->


<!-- ```{r, fig.width=8, fig.height=8} -->
<!-- SCE_topics <- GetTopics(SCE_model, SCE) -->
<!-- heatmap(SCE, SCE_topics, SCE@colData@listData$hash.ID, "Timepoint") -->

<!-- ``` -->

<!-- ```{r, warning=F, fig.width=8, fig.height=6} -->
<!-- SCE_FeaturePlot(SCE, pt.size = 0.01, topics = 1:8, ncol = 4, min.cutoff = 'q5') -->
<!-- SCE_FeaturePlot(SCE, pt.size = 0.01, topics = 9:16, ncol = 4, min.cutoff = 'q5') -->
<!-- ``` -->

<!-- ```{r echo = F, warning=F, fig.width=8, fig.height=3} -->
<!-- SCE_FeaturePlot(SCE, pt.size = 0.01, topics = 17:20, ncol = 4, min.cutoff = 'q5') -->

<!-- ``` -->







